jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java

```diff
@@ -47,13 +47,55 @@ public class AsyncContent implements Content.Sink, Content.Source, Closeable
     private Runnable demandCallback;
     private long length = UNDETERMINED_LENGTH;
 
+    /**
+     * {@inheritDoc}
+     * <p>The write completes when the {@link Content.Chunk} returned by {@link #read()}
+     * that wraps {@code byteBuffer} is released.</p>
+     */
     @Override
     public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
     {
-        write(Content.Chunk.from(byteBuffer, last, callback::succeeded), callback);
+        Content.Chunk chunk;
+        if (!byteBuffer.hasRemaining())
+            chunk = last ? Content.Chunk.EOF : Content.Chunk.EMPTY;
+        else
+            chunk = Content.Chunk.from(byteBuffer, last, callback::succeeded);
+        offer(chunk, callback);
     }
 
+    /**
+     * <p>Writes the given {@link ByteBuffer}, notifying the {@link Callback} when the
+     * write is complete.</p>
+     * <p>The callback completes:</p>
+     * <ul>
+     * <li>immediately with a failure when the written chunk is an instance of {@link Content.Chunk.Error}</li>
```

Comment:
This feel weird.
I think it should be treated like a normal "special" write of, say, `EOF`, so succeeding the callback.